### PR TITLE
Add Chapter Support

### PR DESCRIPTION
**Pull request by Arthi-chaud**
Hello!

This PR is a contribution similar to #45, with the fixes you requested in the PR.

Regarding testing, I added chapters to the `test.mp4` file and added/updated some tests.

Please let me know if something else is needed.

### CODE REVIEW

File: ffprobe_test.go [hunk 1]
```diff
@@ -154,7 +155,7 @@ func validateData(t *testing.T, data *ProbeData) {
 	}
 
 	stream = data.StreamType(StreamAny)
-	if len(stream) != 2 {
+	if len(stream) != 3 {
 		t.Errorf("It should have two streams.")
```
Comment (vansante) [inline ffprobe_test.go]: also adjust the comment please

File: probedata.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ const (
 type ProbeData struct {
 	Streams []*Stream `json:"streams"`
 	Format  *Format   `json:"format"`
+	Chapters []*Chapter  `json:"chapters"`
```
Comment (vansante) [inline probedata.go]: can you run gofmt to fix the spacing?

File: probedata.go [hunk 3]
```diff
@@ -102,6 +103,35 @@ type StreamDisposition struct {
 	AttachedPic     int `json:"attached_pic"`
 }
 
+// Chapters is a json data structure to represent chapters.
+type Chapter struct {
+	ID               int         `json:"id"`
+	TimeBase         string      `json:"time_base"`
+	StartTimeSeconds float64     `json:"start_time,string"`
+	EndTimeSeconds   float64     `json:"end_time,string"`
+	Tags             *ChapterTags `json:"tags"`
```
Comment (vansante) [inline probedata.go]: If this is just a mapping, cant we re-use the `Tags` type, saving us from having to add all props separately?

File: probedata.go [hunk 4]
```diff
@@ -102,6 +103,35 @@ type StreamDisposition struct {
...
+}
+
+// ChapterTags is a json data structure to represent metadata of a chapter.
+type ChapterTags struct {
+	Title string `json:"title"`
+}
+
+// StartTime returns the start time of the chapter as a time.Duration
+func (c *Chapter) StartTime() (duration time.Duration) {
```
Comment (vansante) [inline probedata.go]: I dont think there is a need to declare the return var, so:
```suggestion
func (c *Chapter) StartTime() time.Duration {
```

File: probedata.go [hunk 5]
```diff
@@ -102,6 +103,35 @@ type StreamDisposition struct {
...
+	return time.Duration(c.StartTimeSeconds * float64(time.Second))
+}
+
+// EndTime returns the end timestamp of the chapter as a time.Duration
+func (c *Chapter) EndTime() (duration time.Duration) {
```
Comment (vansante) [inline probedata.go]:
```suggestion
func (c *Chapter) EndTime() time.Duration {
```

**Review by vansante — COMMENTED** [review]
Some comments

**Comment by Arthi-chaud** [comment]
Everything's been fixed! :)

File: probedata.go [hunk 6]
```diff
@@ -102,6 +103,31 @@ type StreamDisposition struct {
...
+	return time.Duration(c.StartTimeSeconds * float64(time.Second))
+}
+
+// EndTime returns the end timestamp of the chapter as a time.Duration
+func (c *Chapter) EndTime() time.Duration {
+	return time.Duration(c.EndTimeSeconds * float64(time.Second))
+}
+
+// Name returns the value of the "title" tag of the chapter
+func (c *Chapter) Name() string {
```
Comment (vansante) [inline probedata.go]: Why name the function `Name` when it returns the title? Isnt it better to name it `Title` then?

File: probedata.go [hunk 7]
```diff
@@ -102,6 +103,31 @@ type StreamDisposition struct {
...
+	title, _ := c.TagList.GetString("title")
```
Comment (vansante) [inline probedata.go:127]: Is this always error-less (safe to ignore)?

**Review by vansante — COMMENTED** [review]
I have a few more nitpicks

File: probedata.go (hunk 7 above)
Comment (Arthi-chaud) [reply probedata.go:127]: Yes, if a chapter does not have a title, returning an empty string seems like the right thing to do.

**Comment by Arthi-chaud** [comment]
Should I do something about the lint errors? (since it's caused by the test)

Also, not sure why the test workflow fails (Tested in a dockerised environment, with the same setup as the workflow). Help would be appreciated. :)

**Comment by vansante** [comment]
The linter fails because the `validateData()` function is now too long(=complex), which means the best course of action is to split it up in `validateStreams()`, `validateChapter()`, etc.

As for the test failure, its fails on:
>     ffprobe_test.go:143: It does not have a subtitle stream.
>     ffprobe_test.go:149: It does not have a data stream.

The second one you adjusted in your test update, was that needed?
As for the first one, I have honestly no clue, will look into it later.

**Comment by Arthi-chaud** [comment]
I split the tests into smaller functions.

It looks like the mp4 containers' handling of chapters adds a stream which, depending on the version of `ffprobe`, can be interpreted as a subtitle stream (posts about it [here](https://trac.ffmpeg.org/ticket/9436) and [here](https://forum.videohelp.com/threads/402064-MP4-chapters-and-data-stream))
On my machine (macOS, M1), I do not have the issue. My guess is that it is related to `ffmpeg` itself.

**Comment by vansante** [comment]
Let me see if I can upgrade some things (like ffprobe) to fix it.

**Comment by vansante** [comment]
Allright, I have upgraded the pipelines to run the latest release version of ffprobe, however, I did not notice any difference in tests:

https://github.com/vansante/go-ffprobe/pull/50

**Comment by vansante** [comment]
I guess I would suggest to cater the test results to what we expect on an ubuntu machine.

**Comment by Arthi-chaud** [comment]
Hi! Just rebased the fork so that it would use the updated GH actions.

[Looks like](https://github.com/vansante/go-ffprobe/actions/runs/9301902399) the 2nd run of the test did not use the latest version of the tests. Would you mind re-running time just to check?

**Review by vansante — APPROVED** [review]
Looks good, lets merge it!

**Comment by vansante** [comment]
It was time for a new release, so [here](https://github.com/vansante/go-ffprobe/releases/tag/v2.2.0) it is

**Pull request merged**